libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1696 @@
+    }
+    
+    // parse_var parses a single type_var
+    fn parse_type_var(& mut self, token: &Token) -> Result<Tvar, &'static str>{
+        match &token.text {
+            Some(text) => { 
+                let num = text.trim_start_matches("t").parse::<i64>();
+                match num {
+                    Err(_e) => Err("Not a valid type variable"), 
+                    Ok (num) => {
+                        self.next();
+                        Ok(Tvar(num))
+                    }
+                }
+            }
+            None => Err("Type variable must have text")
+        }
+    }
+
+    // parse_contraints parses a list of constraints for each type_var with contraints
+    fn parse_constraints(& mut self) -> Option<TvarKinds> {
+        let mut cons_map = HashMap::new();
+
+        loop {
+            let mut next_token = self.peek(); 
+
+            if next_token.token_type == TokenTypes::COMMA {
+                self.next(); // skip to comma
+                next_token = self.peek(); // look at identifier next
+            } 
+
+            if next_token.token_type == TokenTypes::IDENTIFIER {
+                let type_var = self.parse_type_var(&next_token);
+                let kinds = self.parse_kinds(); 
+                match type_var {
+                    Err(_e) => return { None } , 
+                    Ok(var) => 
+                    match kinds {
+                        Err(_e) => return { None } , 
+                        Ok(kinds) => cons_map.insert(var, kinds), 
+                    }
+                };
+            } else {
+                break;
+            }
+        }
+        Some(TvarKinds(cons_map))
+    }
+
+    // parse_kinds parses a list of kinds to associate with a type_var for a constraint
+    fn parse_kinds(& mut self) -> Result<HashSet<Kind>, &'static str>{
+        let mut kinds_set = HashSet::new(); 
+        
+        loop {
+            let next_token = self.peek(); 
+            if next_token.token_type != TokenTypes::COLON {
+                if next_token.token_type != TokenTypes::PLUS {
+                    break; 
+                }  
+            }
+            self.next(); 
+            let kind = self.parse_kind();
+
+            match kind {
+                Err(e) => return Err(e),
+                Ok(kind) => {
+                    kinds_set.insert(kind);
+                } 
+            }
+        }
+        Ok(kinds_set)
+    }
+
+    // parse_kind parses a single kind for a constraint
+    fn parse_kind(& mut self) -> Result<Kind, &'static str> {
+        let token = self.next();
+
+        let addable = "Addable".to_string();
+        let subtractable = "Subtractable".to_string();
+        let divisible = "Divisible".to_string();
+        let comparable = "Comparable".to_string();
+        let nullable = "Nullable".to_string();
+        let equatable = "Equatable".to_string();
+
+        match token.text {
+            Some(text) => {
+                if text == addable {
```

Comment:
Is it possible in Rust to create a map of strings and use that for this? This seems like a candidate for a map of strings to kinds.

Reply:
What if instead of converting this to a map, I got rid of the variables and simplified the match expression? Do you still think a map would be better?

Reply:
I just don't like the idea of creating a whole map and iterating over it when I can do this with a map statement.

Reply:
If a match expression works, go with that. I was thinking it would be a global map and usually you can just access the map. I don't think you need to iterate over it.